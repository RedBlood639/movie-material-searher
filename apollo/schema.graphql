enum Media {
  movie
  tv
}

type UpcomingResults {
  id: Int!
  media_type: Media!
  poster_path: String
  release_date: String!
  title: String!
}

type Upcoming {
  page: Int!
  results: [UpcomingResults!]!
  total_pages: Int!
}

type NowPlayingResults {
  id: Int!
  media_type: Media!
  poster_path: String
  title: String!
  vote_average: Float!
}

type NowPlaying {
  page: Int!
  results: [NowPlayingResults!]!
  total_pages: Int!
}

type MoviesSearchResults {
  id: Int!
  media_type: Media!
  title: String!
}

type MoviesSearch {
  cursor: Int
  hasMore: Boolean
  page: Int!
  results: [MoviesSearchResults!]!
  total_pages: Int!
}

type SimilarResults {
  id: Int!
  media_type: Media!
  poster_path: String
  release_date: String!
  title: String!
}

type SimilarMovies {
  cursor: Int
  hasMore: Boolean
  results: [SimilarResults!]!
}

type MovieInfo {
  backdrop_path: String
  budget: String!
  id: Int!
  media_type: Media!
  overview: String
  poster_path: String
  release_date: String!
  revenue: String!
  similar: SimilarMovies!
  title: String!
  vote_average: Float!
}

type TvShowInfo {
  backdrop_path: String
  id: Int!
  media_type: Media!
  number_of_episodes: Int!
  number_of_seasons: Int!
  overview: String
  poster_path: String
  release_date: String!
  similar: SimilarMovies!
  title: String!
  vote_average: Float!
}

type Query {
  nowPlaying(page: Int = 1): NowPlaying!
  movieInfo(id: ID!, cursor: Int, pageSize: Int = 4): MovieInfo!
  moviesSearch(query: String!, cursor: Int, pageSize: Int = 8): MoviesSearch!
  tvShowInfo(id: ID!, cursor: Int, pageSize: Int = 4): TvShowInfo!
  upcoming(page: Int = 1): Upcoming!
}
