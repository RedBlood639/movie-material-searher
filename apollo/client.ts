import {
  ApolloClient,
  InMemoryCache,
  makeVar,
  NormalizedCacheObject,
} from '@apollo/client'
import merge from 'deepmerge'
import isEqual from 'lodash.isequal'
import { useMemo } from 'react'

export type Favorite = {
  id: number
  media_type: string
  poster_path?: string | null
  title: string
}

export const favoritesVar = makeVar<Favorite[]>([])
export const prefersDarkModeVar = makeVar(false)

let apolloClient: ApolloClient<NormalizedCacheObject>

function createApolloClient() {
  return new ApolloClient({
    cache: new InMemoryCache(),
    ssrMode: typeof window === 'undefined',
    uri: process.env.NEXT_PUBLIC_SERVER_URL,
  })
}

export function initializeApollo(initialState: {} | null = null) {
  const internalApolloClient = apolloClient ?? createApolloClient()

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = internalApolloClient.extract()

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = merge(initialState, existingCache, {
      // combine arrays using object equality (like in sets)
      arrayMerge: (destinationArray, sourceArray) => [
        ...sourceArray,
        ...destinationArray.filter(d => sourceArray.every(s => !isEqual(d, s))),
      ],
    })

    // Restore the cache with the merged data
    internalApolloClient.cache.restore(data)
  }
  // For SSG and SSR always create a new Apollo Client
  if (typeof window === 'undefined') return internalApolloClient
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = internalApolloClient

  return internalApolloClient
}

const APOLLO_STATE_PROP_NAME = '__APOLLO_STATE__'

export function addApolloState(
  client: ApolloClient<NormalizedCacheObject>,
  pageProps: any,
) {
  if (pageProps?.props) {
    // eslint-disable-next-line no-param-reassign
    pageProps.props[APOLLO_STATE_PROP_NAME] = client.cache.extract()
  }

  return pageProps
}

export function useApollo(pageProps: any) {
  const state = pageProps[APOLLO_STATE_PROP_NAME]
  const store = useMemo(() => initializeApollo(state), [state])
  return store
}
