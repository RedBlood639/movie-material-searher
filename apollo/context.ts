import MoviesAPI from './datasource'

export type Context = {
  dataSources: {
    moviesAPI: MoviesAPI
  }
}
