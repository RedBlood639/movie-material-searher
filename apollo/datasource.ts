/* eslint-disable class-methods-use-this */
import { RequestOptions, RESTDataSource } from 'apollo-datasource-rest'
import dayjs from 'dayjs'
import numeral from 'numeral'

import {
  Media,
  MovieInfo,
  MoviesSearch,
  NowPlaying,
  TvShowInfo,
  Upcoming,
} from './__generated__'
import {
  MockMovieInfoResponse,
  MockMoviesSearchResponse,
  MockNowPlayingResponse,
  MockTvShowInfoResponse,
  MockUpcomingResponse,
} from './__responses__/raw-responses'

export default class MoviesAPI extends RESTDataSource {
  constructor() {
    super()
    this.baseURL = 'https://api.themoviedb.org/3'
  }

  protected willSendRequest(request: RequestOptions) {
    request.params.set('api_key', process.env.MOVIE_API_KEY as string)
  }

  private attachPoster(path: string, size = 200) {
    return path ? `https://image.tmdb.org/t/p/w${size}${path}` : null
  }

  private moviesUpcomingReducer(movies: MockUpcomingResponse): Upcoming {
    return {
      page: movies.page,
      results: movies.results.map(movie => ({
        id: movie.id,
        media_type: Media.Movie,
        poster_path: this.attachPoster(movie.poster_path),
        release_date: dayjs(movie.release_date).format('DD.MM.YYYY'),
        title: movie.title,
      })),
      total_pages: movies.total_pages,
    }
  }

  private moviesNowPlayingReducer(movies: MockNowPlayingResponse): NowPlaying {
    return {
      page: movies.page,
      results: movies.results.map(movie => ({
        id: movie.id,
        media_type: Media.Movie,
        poster_path: this.attachPoster(movie.poster_path),
        title: movie.title,
        vote_average: movie.vote_average,
      })),
      total_pages: movies.total_pages,
    }
  }

  private moviesSearchReducer(movies: MockMoviesSearchResponse): MoviesSearch {
    return {
      page: movies.page,
      results: movies.results
        .filter(movie => movie.media_type !== 'person')
        .map(movie => ({
          id: movie.id,
          media_type: movie.media_type as Media,
          title: movie?.title || movie?.name || '',
        })),
      total_pages: movies.total_pages,
    }
  }

  private movieInfoReducer(movie: MockMovieInfoResponse): MovieInfo {
    return {
      backdrop_path: this.attachPoster(movie.backdrop_path, 500),
      budget: numeral(movie.budget).format('$0,00'),
      id: movie.id,
      media_type: Media.Movie,
      overview: movie.overview,
      poster_path: this.attachPoster(movie.poster_path),
      release_date: movie.release_date
        ? dayjs(movie.release_date).format('DD MMMM YYYY')
        : '',
      revenue: numeral(movie.revenue).format('$0,00'),
      similar: {
        results: movie.similar.results.map(similarMovie => ({
          id: similarMovie.id,
          media_type: Media.Movie,
          poster_path: this.attachPoster(similarMovie.poster_path),
          release_date: similarMovie.release_date
            ? dayjs(similarMovie.release_date).format('YYYY')
            : '',
          title: similarMovie.title,
        })),
      },
      title: movie.title,
      vote_average: movie.vote_average,
    }
  }

  private tvShowInfoReducer(tv: MockTvShowInfoResponse): TvShowInfo {
    return {
      backdrop_path: this.attachPoster(tv.backdrop_path, 500),
      id: tv.id,
      media_type: Media.Tv,
      number_of_episodes: tv.number_of_episodes,
      number_of_seasons: tv.number_of_seasons,
      overview: tv.overview,
      poster_path: this.attachPoster(tv.poster_path),
      release_date: tv.first_air_date
        ? dayjs(tv.first_air_date).format('DD MMMM YYYY')
        : '',
      similar: {
        results: tv.similar.results.map(similarTvShow => ({
          id: similarTvShow.id,
          media_type: Media.Tv,
          poster_path: this.attachPoster(similarTvShow.poster_path),
          release_date: similarTvShow.first_air_date
            ? dayjs(similarTvShow.first_air_date).format('YYYY')
            : '',
          title: similarTvShow.name,
        })),
      },
      title: tv.name,
      vote_average: tv.vote_average,
    }
  }

  async getUpcoming(page: number) {
    const data: MockUpcomingResponse = await this.get('/movie/upcoming', {
      page,
    })
    return this.moviesUpcomingReducer(data)
  }

  async getNowPlaying(page: number) {
    const data: MockNowPlayingResponse = await this.get('/movie/now_playing', {
      page,
    })
    return this.moviesNowPlayingReducer(data)
  }

  async getMoviesSearch(query: string, page: number) {
    const data: MockMoviesSearchResponse = await this.get('/search/multi', {
      page,
      query,
    })
    return this.moviesSearchReducer(data)
  }

  async getMovieInfo(id: string) {
    const data: MockMovieInfoResponse = await this.get(`/movie/${id}`, {
      append_to_response: 'similar',
    })
    return this.movieInfoReducer(data)
  }

  async getTvShowInfo(id: string) {
    const data: MockTvShowInfoResponse = await this.get(`/tv/${id}`, {
      append_to_response: 'similar',
    })
    return this.tvShowInfoReducer(data)
  }
}
