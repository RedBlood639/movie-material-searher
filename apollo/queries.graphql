query GetMovieInfo($id: ID!, $pageSize: Int) {
  movieInfo(id: $id, pageSize: $pageSize) {
    backdrop_path
    budget
    id
    media_type
    overview
    poster_path
    release_date
    revenue
    similar {
      results {
        id
        media_type
        poster_path
        release_date
        title
      }
    }
    title
    vote_average
  }
}

query GetTvShowInfo($id: ID!, $pageSize: Int) {
  tvShowInfo(id: $id, pageSize: $pageSize) {
    backdrop_path
    id
    media_type
    number_of_episodes
    number_of_seasons
    overview
    poster_path
    release_date
    similar {
      results {
        id
        media_type
        poster_path
        release_date
        title
      }
    }
    title
    vote_average
  }
}

query GetMovies {
  nowPlaying {
    page
    results {
      id
      media_type
      poster_path
      title
      vote_average
    }
    total_pages
  }
  upcoming {
    page
    results {
      id
      media_type
      poster_path
      release_date
      title
    }
    total_pages
  }
}

query GetUpcoming($page: Int) {
  upcoming(page: $page) {
    page
    results {
      id
      media_type
      poster_path
      release_date
      title
    }
    total_pages
  }
}

query GetNowPlaying($page: Int) {
  nowPlaying(page: $page) {
    page
    results {
      media_type
      id
      poster_path
      title
      vote_average
    }
    total_pages
  }
}

query GetMoviesSearch($query: String!, $pageSize: Int) {
  moviesSearch(query: $query, pageSize: $pageSize) {
    results {
      id
      media_type
      title
    }
  }
}
