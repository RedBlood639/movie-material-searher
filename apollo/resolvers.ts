import { MoviesSearch, Resolvers, SimilarMovies } from './__generated__'

function paginateResults<T extends { results: { [key: string]: any } }>(
  data: T,
  size: number,
  cursor?: number | null,
) {
  let results: []

  if (!cursor) {
    results = data.results.slice(0, size)
  } else {
    results = data.results.slice(0, cursor + size)
  }

  return {
    ...data,
    cursor: results.length,
    hasMore: results.length !== data.results.length,
    results,
  }
}

const resolvers: Resolvers = {
  Query: {
    async movieInfo(_root, { cursor, id, pageSize }, { dataSources }) {
      try {
        const data = await dataSources.moviesAPI.getMovieInfo(id)
        return {
          ...data,
          similar: paginateResults<SimilarMovies>(
            data.similar,
            pageSize,
            cursor,
          ),
        }
      } catch (error) {
        throw new Error(error)
      }
    },
    async moviesSearch(_root, { cursor, pageSize, query }, { dataSources }) {
      const fetchMovies = async (page = 1): Promise<MoviesSearch> => {
        const data = await dataSources.moviesAPI.getMoviesSearch(query, page)

        if (!data.results.length && data.total_pages > 1) {
          return fetchMovies(data.page + 1)
        }

        return data
      }

      try {
        const data = await fetchMovies()
        return paginateResults<MoviesSearch>(data, pageSize, cursor)
      } catch (error) {
        throw new Error(error)
      }
    },
    async nowPlaying(_root, { page }, { dataSources }) {
      try {
        const data = await dataSources.moviesAPI.getNowPlaying(page)

        if (!data.results.length || data.page !== +page) {
          throw new Error('404 Not found')
        }

        return data
      } catch (error) {
        throw new Error(error)
      }
    },
    async tvShowInfo(_root, { cursor, id, pageSize }, { dataSources }) {
      try {
        const data = await dataSources.moviesAPI.getTvShowInfo(id)
        return {
          ...data,
          similar: paginateResults<SimilarMovies>(
            data.similar,
            pageSize,
            cursor,
          ),
        }
      } catch (error) {
        throw new Error(error)
      }
    },
    async upcoming(_root, { page }, { dataSources }) {
      try {
        const data = await dataSources.moviesAPI.getUpcoming(page)

        if (!data.results.length || data.page !== +page) {
          throw new Error('404 Not found')
        }

        return data
      } catch (error) {
        throw new Error(error)
      }
    },
  },
}

export default resolvers
