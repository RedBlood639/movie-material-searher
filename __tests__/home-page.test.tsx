import { GetMoviesDocument } from '@/apollo/__generated__'
import { renderApollo, screen } from '@/lib/setupTests'
import HomePage from '@/pages'

const mocks = [
  {
    request: {
      query: GetMoviesDocument,
    },
    result: {
      data: {
        nowPlaying: {
          page: 1,
          results: [
            {
              __typename: 'NowPlayingResults',
              id: 1,
              media_type: 'movie',
              poster_path: 'https://example.com/img/1',
              title: 'now playing rendered',
              vote_average: 5,
            },
          ],
          total_pages: 20,
        },
        upcoming: {
          page: 1,
          results: [
            {
              __typename: 'UpcomingResults',
              id: 1,
              media_type: 'movie',
              poster_path: null,
              release_date: '2002',
              title: 'upcoming rendered',
            },
          ],
          total_pages: 20,
        },
      },
    },
  },
]

describe('home page', () => {
  it('should fetch initial movies', async () => {
    const { baseElement } = renderApollo(<HomePage />, {
      addTypeName: true,
      mocks,
    })

    expect(await screen.findByText('upcoming rendered')).toBeInTheDocument()
    expect(await screen.findByText('now playing rendered')).toBeInTheDocument()
    expect(baseElement).toMatchSnapshot()
  })

  it('should render error state', async () => {
    const mock = [
      {
        error: new Error('an error has occurred'),
        request: {
          query: GetMoviesDocument,
        },
      },
    ]

    renderApollo(<HomePage />, { mocks: mock })

    expect(
      await screen.findByText(/an error has occurred/i),
    ).toBeInTheDocument()
  })
})
