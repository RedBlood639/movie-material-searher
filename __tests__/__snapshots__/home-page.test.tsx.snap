// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home page should fetch initial movies 1`] = `
<body>
  <div>
    <div
      class="MuiGrid-root MuiGrid-container css-15fev7r-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-6 css-rdfm1p-MuiGrid-root"
      >
        <div
          class="css-1ry038z-Container emm7brd0"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation10 efhkdvm1 css-103l67h-MuiPaper-root-Container"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter MuiTypography-gutterBottom css-1trk376-MuiTypography-root"
            >
              Now Playing
            </h4>
            <div
              class="MuiGrid-root MuiGrid-container css-15fev7r-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-3 efhkdvm0 css-11ine10-MuiGrid-root-Wrapper"
              >
                <div
                  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root ejzktm53 css-18ogkrw-MuiPaper-root-MuiCard-root-CardContainer"
                >
                  <button
                    class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
                    tabindex="0"
                    type="button"
                  >
                    <div
                      style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
                    >
                      <div
                        style="box-sizing: border-box; display: block; max-width: 100%;"
                      >
                        <img
                          alt=""
                          aria-hidden="true"
                          role="presentation"
                          src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjAwIiBoZWlnaHQ9IjMwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
                          style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
                        />
                      </div>
                      <img
                        alt="now playing rendered"
                        decoding="async"
                        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                        style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
                      />
                    </div>
                    <div
                      class="MuiCardContent-root ejzktm52 css-dzh1mx-MuiCardContent-root-StyledCardContent"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-inherit ejzktm50 css-eh8d4t-MuiTypography-root-StyledTypography"
                      >
                        now playing rendered
                      </p>
                      <p
                        class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium ejzktm51 css-1uoctnc-MuiSvgIcon-root-StyledStarIcon"
                          data-testid="StarIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                          />
                        </svg>
                        5
                      </p>
                    </div>
                    <span
                      class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
          <nav
            aria-label="pagination navigation"
            class="MuiPagination-root MuiPagination-text e10om17h0 css-ue6cel-MuiPagination-root-StyledPagination"
          >
            <ul
              class="MuiPagination-ul css-wjh20t-MuiPagination-ul"
            >
              <li>
                <button
                  aria-label="Go to previous page"
                  class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-page css-12h979j-MuiButtonBase-root-MuiPaginationItem-root"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                    data-testid="NavigateBeforeIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                    />
                  </svg>
                </button>
              </li>
              <li>
                <button
                  aria-current="true"
                  aria-label="page 1"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-selected MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  1
                </button>
              </li>
              <li>
                <button
                  aria-label="Go to page 2"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  2
                </button>
              </li>
              <li>
                <button
                  aria-label="Go to page 3"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  3
                </button>
              </li>
              <li>
                <button
                  aria-label="Go to page 4"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  4
                </button>
              </li>
              <li>
                <button
                  aria-label="Go to page 5"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  5
                </button>
              </li>
              <li>
                <div
                  aria-label="Go to end-ellipsis page"
                  class="MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-ellipsis css-1v2lvtn-MuiPaginationItem-root"
                >
                  â€¦
                </div>
              </li>
              <li>
                <button
                  aria-label="Go to page 20"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  20
                </button>
              </li>
              <li>
                <button
                  aria-label="Go to next page"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-page css-12h979j-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                    data-testid="NavigateNextIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                    />
                  </svg>
                </button>
              </li>
            </ul>
          </nav>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-6 css-rdfm1p-MuiGrid-root"
      >
        <div
          class="css-1ry038z-Container emm7brd0"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation10 efhkdvm1 css-103l67h-MuiPaper-root-Container"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter MuiTypography-gutterBottom css-1trk376-MuiTypography-root"
            >
              Upcoming
            </h4>
            <div
              class="MuiGrid-root MuiGrid-container css-15fev7r-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-3 efhkdvm0 css-11ine10-MuiGrid-root-Wrapper"
              >
                <div
                  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root ejzktm53 css-18ogkrw-MuiPaper-root-MuiCard-root-CardContainer"
                >
                  <button
                    class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
                    tabindex="0"
                    type="button"
                  >
                    <div
                      style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
                    >
                      <div
                        style="box-sizing: border-box; display: block; max-width: 100%;"
                      >
                        <img
                          alt=""
                          aria-hidden="true"
                          role="presentation"
                          src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjAwIiBoZWlnaHQ9IjMwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
                          style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
                        />
                      </div>
                      <img
                        alt="upcoming rendered"
                        decoding="async"
                        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                        style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
                      />
                    </div>
                    <div
                      class="MuiCardContent-root ejzktm52 css-dzh1mx-MuiCardContent-root-StyledCardContent"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-inherit ejzktm50 css-eh8d4t-MuiTypography-root-StyledTypography"
                      >
                        upcoming rendered
                      </p>
                      <p
                        class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
                      >
                        2002
                      </p>
                    </div>
                    <span
                      class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
          <nav
            aria-label="pagination navigation"
            class="MuiPagination-root MuiPagination-text e10om17h0 css-ue6cel-MuiPagination-root-StyledPagination"
          >
            <ul
              class="MuiPagination-ul css-wjh20t-MuiPagination-ul"
            >
              <li>
                <button
                  aria-label="Go to previous page"
                  class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-page css-12h979j-MuiButtonBase-root-MuiPaginationItem-root"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                    data-testid="NavigateBeforeIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                    />
                  </svg>
                </button>
              </li>
              <li>
                <button
                  aria-current="true"
                  aria-label="page 1"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-selected MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  1
                </button>
              </li>
              <li>
                <button
                  aria-label="Go to page 2"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  2
                </button>
              </li>
              <li>
                <button
                  aria-label="Go to page 3"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  3
                </button>
              </li>
              <li>
                <button
                  aria-label="Go to page 4"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  4
                </button>
              </li>
              <li>
                <button
                  aria-label="Go to page 5"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  5
                </button>
              </li>
              <li>
                <div
                  aria-label="Go to end-ellipsis page"
                  class="MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-ellipsis css-1v2lvtn-MuiPaginationItem-root"
                >
                  â€¦
                </div>
              </li>
              <li>
                <button
                  aria-label="Go to page 20"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  20
                </button>
              </li>
              <li>
                <button
                  aria-label="Go to next page"
                  class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-page css-12h979j-MuiButtonBase-root-MuiPaginationItem-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                    data-testid="NavigateNextIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                    />
                  </svg>
                </button>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </div>
  </div>
</body>
`;
